### PR TITLE
Improved Test Coverage and Robustness for QuestRepositoryFireStore

### PR DESCRIPTION
**Pull request by ziyadbagh**
#### Pull Request Description:
This pull request introduces major enhancements and increased test coverage for the `QuestRepositoryFireStore` class.

#### Summary of Changes:

1. **Increased Test Coverage**:
   - Achieved a coverage increase from 62% to 86% by adding targeted tests for all branches and handling various edge cases in `getDailyQuest`.
   - Expanded `documentToQuest` tests to confirm correct behavior when fields are missing, returning `null`, or causing exceptions.

2. **Comprehensive Tests for `getDailyQuest`**:
   - **Success Scenarios**: Added tests to confirm that quests are consistently sorted in ascending order by index, covering cases like non-sequential indices, duplicate values, and negative values.
   - **Failure Scenarios**: Simulated multiple failure conditions, including an empty document list, `null` task results, and Firestore errors.
   - **Firestore Mocks and Captors**: Used mocks and captors for `OnCompleteListener` to realistically simulate Firestore’s task completion, ensuring test cases mirror actual interactions closely.

3. **Enhanced Initialization Logic**:
   - Verified the `init` method to confirm it behaves appropriately based on user login status, ensuring the `onSuccess` callback triggers only when a user is logged in, thus strengthening app security and initialization accuracy.
   
image

### CODE REVIEW

**Review by NBCCF025 — APPROVED** [review]
This PR offers valuable improvements in both the functionality and test coverage of the QuestRepositoryFireStore class. The increase in coverage from 61% to 85% is impressive and demonstrates a strong focus on reliability.

Ensure the test function name is in a single-word format to maintain consistency with the naming conventions used in other test files. This alignment will improve readability and help maintain uniformity across the codebase.

Overall, this PR represents a well-structured and valuable enhancement to the QuestRepositoryFireStore class, focusing on reliability, error handling, and user security. I approve this PR, great job!

File: app/src/test/java/com/github/se/signify/model/quest/QuestRepositoryFirestoreTest.kt [hunk 1]
```diff
@@ -89,52 +95,30 @@ class QuestRepositoryFirestoreTest {
...
-    // Arrange
-    val mockDocumentSnapshot1: DocumentSnapshot = mock(DocumentSnapshot::class.java)
-    `when`(mockDocumentSnapshot1.getString("title")).thenReturn("Quest Title 1")
-    `when`(mockDocumentSnapshot1.getString("description")).thenReturn("Quest Description 1")
-    `when`(mockDocumentSnapshot1.getString("index")).thenReturn("1")
+  fun init_doesNotTriggerOnSuccessWhenUserIsNotLoggedIn() {
+    // Mock the static Firebase.auth to return our mockAuth instance
+    mockStatic(FirebaseAuth::class.java).use { firebaseAuthMock ->
+      firebaseAuthMock.`when`<FirebaseAuth> { FirebaseAuth.getInstance() }.thenReturn(mockAuth)
+      `when`(mockAuth.currentUser).thenReturn(null) // Simulate user not logged in
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/quest/QuestRepositoryFirestoreTest.kt:102]: check edge cases for the case when the user is not logged in the app.

File: app/src/test/java/com/github/se/signify/model/quest/QuestRepositoryFirestoreTest.kt [hunk 2]
```diff
@@ -165,143 +173,116 @@ class QuestRepositoryFirestoreTest {
...
+    val quest = questRepositoryFirestore.documentToQuest(mockDocument)
+
+    // Assert
+    assertNull(quest)
   }
 
   @Test
-  fun getDailyQuest_successfulFetch_callsOnSuccessWithQuests() {
-    // Arrange: Mock DocumentSnapshots
+  fun getDailyQuest_successfulFetch_returnsSortedQuests() {
```
Comment (SaFouNaldo) [inline app/src/test/java/com/github/se/signify/model/quest/QuestRepositoryFirestoreTest.kt:189]: good joob on checking that the order is maintained

**Review by SaFouNaldo — APPROVED** [review]
- Reviewed edge case handling for questFireStore, and these checks significantly enhance test coverage and robustness. Good attention to detail!
- Code cleanup makes the logic more readable and maintainable. The refactoring improves overall code quality without altering functionality.

Great work on improving both functionality and readability!
